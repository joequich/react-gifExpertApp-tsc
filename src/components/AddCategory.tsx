import React, { ChangeEvent, Dispatch, FormEvent, SetStateAction, useState } from 'react';
import PropTypes from 'prop-types';

interface Props {
    setCategories: Dispatch<SetStateAction<string[]>>;
}

export const AddCategory = ({setCategories}: Props) => {
    const [inputValue, setInputValue] = useState<string>('');

    const handleInputChange = (e: ChangeEvent<HTMLInputElement>) => {
        setInputValue(e.target.value);
    }

    const handleSubmit = (e: FormEvent<HTMLElement>) => {
        e.preventDefault();

        if (inputValue.trim().length > 2) {
            setCategories(cats => {
                const exist = cats.find(cat => cat === inputValue);
                if (exist) return [...cats]; 
                return [inputValue, ...cats]
            });
            setInputValue('');
        }
    }

    return (
        <>
            <form role="search" aria-label="Search a gif" onSubmit={handleSubmit}>
                <input type="text" placeholder="Search" value={inputValue} onChange={handleInputChange}/>
            </form>
        </>
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired,
}