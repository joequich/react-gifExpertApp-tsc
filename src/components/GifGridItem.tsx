import React from 'react'
import PropTypes from 'prop-types';

//IGifs
export interface Props {
    title: string;
    url: string;
}

const GifGridItem = ({title, url}:Props) => {
    return (
        <div role="listitem" className="card animate__animated animate__fadeInUp">
           <img src={url} alt={title} />
            <p>{title}</p>
        </div>
    )
}

GifGridItem.propTypes = {
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
};
GifGridItem.defaultProps = {
    id: ''
};

export default GifGridItem;