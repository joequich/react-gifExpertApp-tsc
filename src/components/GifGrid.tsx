import React from 'react';
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';
import PropTypes from 'prop-types';

type Props = {
    category: string;
}

interface IGifs {
    id: string;
    title: string;
    url: string;
}

export const GifGrid = ({category}:Props) => {
    const { data: images, loading} = useFetchGifs(category);

    return (
        <>
            <h3 className="category-title">{ category }</h3>
            { loading && <p className="animate__pulse">Loading...</p> }
            <div className="card-grid animate__slideInUp">
                {images.map( (img:IGifs) => (
                    <GifGridItem
                    key={img.id}
                    {...img}
                    />
                ))}
            </div>
        </>
    )
}

GifGrid.propType = {
    category: PropTypes.string.isRequired
}
