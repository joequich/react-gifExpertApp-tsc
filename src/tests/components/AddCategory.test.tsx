import { render, screen } from '@testing-library/react';
import { AddCategory } from '../../components/AddCategory';

describe('Tests in <AddCategory />', () => {
    const setCategories = jest.fn();

    const component = render(<AddCategory setCategories={setCategories} />)

    test('should display the component correctly', () => {
        const { asFragment } = component;
        expect(asFragment()).toMatchSnapshot();
    });
});