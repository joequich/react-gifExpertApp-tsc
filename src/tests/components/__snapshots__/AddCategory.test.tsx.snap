// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <AddCategory /> should display the component correctly 1`] = `
<DocumentFragment>
  <form
    aria-label="Search a gif"
    role="search"
  >
    <input
      placeholder="Search"
      type="text"
      value=""
    />
  </form>
</DocumentFragment>
`;
